dipy/segment/metricspeed.pyx

```diff
@@ -0,0 +1,372 @@
+        return same_shape(shape1, shape2)
+
+
+cdef class AveragePointwiseEuclideanMetric(SumPointwiseEuclideanMetric):
+    r""" Computes the average of pointwise Euclidean distances between two
+    sequential data.
+
+    A sequence of N-dimensional points is represented as a 2D array with
+    shape (nb_points, nb_dimensions). A `feature` object can be specified
+    in order to calculate the distance between the features, rather than
+    directly between the sequential data.
+
+    Parameters
+    ----------
+    feature : `Feature` object, optional
+        It is used to extract features before computing the distance.
+
+    Notes
+    -----
+    The distance between two 2D sequential data::
+
+        s1       s2
+
+        0*   a    *0
+          \       |
+           \      |
+           1*     |
+            |  b  *1
+            |      \
+            2*      \
+                c    *2
+
+    is equal to $(a+b+c)/3$ where $a$ is the Euclidean distance between s1[0] and
+    s2[0], $b$ between s1[1] and s2[1] and $c$ between s1[2] and s2[2].
+    """
+    cdef double c_dist(AveragePointwiseEuclideanMetric self, Data2D features1, Data2D features2) nogil:
+        cdef int N = features1.shape[0]
+        cdef double dist = SumPointwiseEuclideanMetric.c_dist(self, features1, features2)
+        return dist / N
+
+
+cdef class MinimumAverageDirectFlipMetric(AveragePointwiseEuclideanMetric):
+    r""" Computes the ADF distance (minimum average direct-flip) between two
+    sequential data.
+
+    A sequence of N-dimensional points is represented as a 2D array with
+    shape (nb_points, nb_dimensions).
+
+    Notes
+    -----
+    The distance between two 2D sequential data::
+
+        s1       s2
+
+        0*   a    *0
+          \       |
+           \      |
+           1*     |
+            |  b  *1
+            |      \
+            2*      \
+                c    *2
+
+    is equal to $\min((a+b+c)/3, (a'+b'+c')/3)$ where $a$ is the Euclidean distance
+    between s1[0] and s2[0], $b$ between s1[1] and s2[1], $c$ between s1[2]
+    and s2[2], $a'$ between s1[0] and s2[2], $b'$ between s1[1] and s2[1]
+    and $c'$ between s1[2] and s2[0].
+    """
+    def __init__(MinimumAverageDirectFlipMetric self):
+        super(MinimumAverageDirectFlipMetric, self).__init__(feature=IdentityFeature())
+
+    property is_order_invariant:
+        """ Is this metric invariant to the sequence's ordering """
+        def __get__(MinimumAverageDirectFlipMetric self):
+            return True  # Ordering is handled in the distance computation
+
+    cdef double c_dist(MinimumAverageDirectFlipMetric self, Data2D features1, Data2D features2) nogil:
+        cdef double dist_direct = AveragePointwiseEuclideanMetric.c_dist(self, features1, features2)
+        cdef double dist_flipped = AveragePointwiseEuclideanMetric.c_dist(self, features1, features2[::-1])
```

Comment:
No, but I'm not really using that metric anyway. When you specify `metric="MDF"`, it uses internally the `AveragePointwiseEuclideanMetric` as the metric with the `IdentityFeature`. Since `IdentityFeature` is not order invariant, so will be `AveragePointwiseEuclideanMetric`. 

In QuickBundles algorithm, two distances are computed (direct-flip) when a metric is not order invariant. That way we can check which of the direct or flip streamlines is given the minimum distance and update the cluster's centroid accordingly.